cpp/src/lists/copying/segmented_gather.cu

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <cudf/detail/gather.hpp>
+#include <cudf/detail/indexalator.cuh>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/lists/detail/gather.cuh>
+
+#include <thrust/binary_search.h>
+#include <rmm/cuda_stream_view.hpp>
+
+namespace cudf {
+namespace lists {
+namespace detail {
+
+std::unique_ptr<column> segmented_gather(lists_column_view const& value_column,
+                                         lists_column_view const& gather_map,
+                                         rmm::cuda_stream_view stream,
+                                         rmm::mr::device_memory_resource* mr)
+{
+  CUDF_EXPECTS(is_index_type(gather_map.child().type()),
+               "Gather map should be list column of index type");
+  CUDF_EXPECTS(!gather_map.has_nulls(), "Gather map contains nulls");
+  CUDF_EXPECTS(value_column.size() == gather_map.size(),
+               "Gather map and list column should be same size");
+
+  // Flattened gather indices
+  auto const gather_map_size = gather_map.get_sliced_child(stream).size();
+  auto child_gather_index =
+    make_numeric_column(data_type{type_to_id<size_type>()}, gather_map_size);
+  auto child_gather_index_begin = child_gather_index->mutable_view().begin<size_type>();
+  auto child_gather_index_end   = child_gather_index->mutable_view().end<size_type>();
+
+  thrust::upper_bound(rmm::exec_policy(stream),
```

Comment:
This wouldn't work if the gather_map was sliced (which is probably unlikely, but you never know). Similar to below, you need to start at the index indicated by 
`gather_map.offsets().begin<size_type>() + gather_map.offset()` 
and subtract 
`(gather_map.offsets().begin<size_type>() + gather_map.offset())[0]`  
from each value

Reply:
Or alternately, throw an exception if someone passes a gather_map with a positive .offset()

Reply:
added sliced gather_map support.